homeassistant/components/bluesound/media_player.py

```diff
@@ -127,6 +125,11 @@ def _stop_polling():
     @callback
     def _add_player_cb():
         """Add player after first sync fetch."""
+        if player.id in [x.id for x in hass.data[DATA_BLUESOUND]]:
```

Comment:
This is ok for now. For the future we should preferably refactor the integration to not store the entities in hass.data but just store a set of known ids for added players. Also refactor the custom service registration to use our entity service helper. And also use our standard join / unjoin media player interface.

https://developers.home-assistant.io/docs/dev_101_services#entity-services

https://developers.home-assistant.io/docs/core/entity/media-player#grouping-player-entities-together

Reply:
And of course, this integration will need a 3rd party library for the device interface for us to accept any major changes.

Reply:
Yeah, I've been planning for two years to refactor this into a seperate library. Hopefully I can find the time for it soon enough. They have also released documentation of the API which will make it easier to make a library that meets my own requirements :)